src/ObjCRuntime/Class.cs

```diff
@@ -91,7 +95,33 @@ public static IntPtr GetHandle (string name)
 		}
 
 		public static IntPtr GetHandle (Type type) {
-			return Register (type);
+			return GetClassHandle (type);
+		}
+
+		static IntPtr GetClassHandle (Type type)
+		{
+			IntPtr @class = IntPtr.Zero;
+
+			// We cache results in a dictionary (type_to_class) - we put failures (when @class = IntPtr.Zero) in the dictionary as well.
+			// We do as little as possible with the lock held (only fetch/add to the dictionary, nothing else)
+
+			bool found;
+			lock (type_to_class)
+				found = type_to_class.TryGetValue (type, out @class);
+
+			if (!found) {
+				@class = FindClass (type, out var is_custom_type);
+				lock (type_to_class)
+					type_to_class [type] = @class;
+			}
+
+			if (@class == IntPtr.Zero) {
+				@class = Register (type);
+				lock (type_to_class)
+					type_to_class [type] = @class;
```

Comment:
how likely is `@class` to remain `null` ? worth checking before the lock ?

Reply:
I put `@class` in the dict even if it's `IntPtr.Zero`, so that the next time we don't have to compute the whole thing just to come up empty once again.